File: csharp/linked-list/LinkedList.cs
using System;
using System.Collections.Generic;

public class Deque<T>
{
    private class Node
    {
        public T Data { get; init; }
        public Node Prev { get; set; }
        public Node Next { get; set; }
    }
    private int _count = 0;
    private Node _head = null;
    private Node _tail = null;

    public Deque() { }
    public Deque(IEnumerable<T> values)
    {
        foreach (T value in values) { Push(value); }
    }
    public void Push(T value)
    {
        Node e = new Node { Data = value };
        if (_count == 0)
        {
            _head = e;
            _tail = e;
        }
        else
        {
            _tail.Next = e;
            e.Prev = _tail;
            _tail = e;
        }
        _count++;
    }

    public T Pop()
    {
        if (_count == 0) { throw new Exception("empty list"); }
        T ret = _tail.Data;
        if (_tail.Prev != null)
        {
            _tail = _tail.Prev;
            _tail.Next = null;
        } else
        {
            _head = null;
            _tail = null;
        }
        _count--;
        return ret;
    }

    public void Unshift(T value)
    {
        Node e = new Node { Data = value };
        if (_count == 0)
        {
            _head = e;
            _tail = e;
        }
        else
        {
            e.Next = _head;
            _head.Prev = e;
            _head = e;
        }
        _count++;
    }

    public T Shift()
    {
        if (_count == 0) { throw new Exception("empty list"); }
        T ret = _head.Data;
        if (_head.Next != null)
        {
            _head = _head.Next;
            _head.Prev = null;
        } else
        {
            _head = null;
            _tail = null;
        }
        _count--;
        return ret;
    }
    // named iterators
    public IEnumerable<T> GetInOrder()
    {
        Node actual = _head;
        while (actual != null)
        {
            yield return actual.Data;
            actual = actual.Next;
        }
    }
    public IEnumerable<T> GetReverseOrder()
    {
        Node actual = _tail;
        while (actual != null)
        {
            yield return actual.Data;
            actual = actual.Prev;
        }
    }
}

File: fsharp/linked-list/LinkedList.fs
module LinkedList

type Node = {
    Data : int 
    mutable Next : Node option 
    mutable Prev : Node option 
}

type LinkedList = {
    mutable Head : Node option 
    mutable Tail : Node option 
}

let mkLinkedList () = { Head = None; Tail = None }

let pop linkedList = 
    match linkedList.Tail with
        | None -> failwith (invalidOp "empty list")
        | Some tailNode ->
            let data = tailNode.Data
            match tailNode.Prev with 
                | None -> 
                    linkedList.Tail <- None
                    linkedList.Head <- None 
                | Some newTail ->
                    newTail.Next <- None 
                    linkedList.Tail <- Some newTail
            data

let shift linkedList =
    match linkedList.Head with
        | None -> failwith (invalidOp "empty list")
        | Some headNode ->
            let data = headNode.Data
            match headNode.Next with 
                | None -> 
                    linkedList.Head <- None
                    linkedList.Tail <- None 
                | Some newHead ->
                    newHead.Prev <- None 
                    linkedList.Head <- Some newHead
            data

let push newValue linkedList = 
    match linkedList.Tail with 
        | None ->
            let newNode = { Data = newValue; Next = None; Prev = None }
            linkedList.Head <- Some newNode 
            linkedList.Tail <- Some newNode
        | Some oldTail -> 
            let newNode = { Data = newValue; Next = None; Prev = Some oldTail }
            oldTail.Next <- Some newNode
            linkedList.Tail <- Some newNode 

let unshift newValue linkedList =
    match linkedList.Head with 
        | None ->
            let newNode = { Data = newValue; Next = None; Prev = None }
            linkedList.Head <- Some newNode 
            linkedList.Tail <- Some newNode
        | Some oldHead -> 
            let newNode = { Data = newValue; Next = Some oldHead; Prev = None }
            oldHead.Prev <- Some newNode
            linkedList.Head <- Some newNode 
